text/0745-user-provided-cdk-init-templates.md

```diff
@@ -0,0 +1,1678 @@
+                * Can slow down the template authoring and review process.
+    5. Template authors develop and maintain their own placeholder and template configuration
+        * Pros:
+            1. Simpler from a user point of view because valid CLI arguments are more concrete
+            2. cdk init does not have to support template configuration
+                * If template authors want to have variations on templates, they can do the branching generation-side, which allows them to use more advanced tools
+        * Cons:
+            1. No default placeholder or template configuration style for template authors
+
+
+
+### What are the drawbacks of this solution?
+
+1. Template validation burden is on template author, can't ensure template quality
+2. Current proposal only supports static templates and not dynamic templates generated from user input
+
+
+
+### What is the high-level project plan?
+
+* Phase 1: Basic support
+
+    1. Support for custom templates through local files, Git repositories, and NPM packages
+    2. Basic CDK supported language type validation and error handling
+
+* Phase 2: Multi-language support
+
+    1. Extend support for C#, F#, Go, Java, JavaScript, and Python
+    2. Develop custom template definition and implement more concrete validation based on that
+
+* Phase 3: Support for additional methods of passing in custom templates
+
+    1. Support for passing in all types of NPM packages such as:
+        1. a folder containing a program described by a package.json file
+        2. a gzipped tarball containing (i)
+        3. a url that resolves to (ii)
+        4. a <name>@<version> that is published on the registry (see registry) with (iii)
+        5. a <name>@<tag> (see npm dist-tag) that points to (iv)
+        6. a <name> that has a "latest" tag satisfying (v)
+        7. a <git remote url> that resolves to (i)
+    2. If a customer wants to use an internal company repository or package manager (GitFarm, CodeArtifact, etc)
+    3. Add CLI parameter for users to specify what branch/commit of repo or package version they want to pull templates from and if not specified then will use main branch/most recent commit or latest package version by default
+
+* Phase 4: Documentation, testing, and marketing
+    * Create documentation for template authors
+    * Create documentation for CDK users
+    * Develop unit and integration tests
+    * Author blog post and other marketing materials
+        * Benefit and example use cases of passing in custom templates through Local/Git/NPM
+        * Benefit and example use cases of selecting branch/commit/package version of a custom template
+
+### Are there any open issues that need to be addressed later?
+
+1. Publish sample templates on GitHub or NPM for template authors to reference when creating
+2. Develop community template registry (initially in CLI and manually updated by CDK team)
```

Comment:
We spoke about having static discovery as part of the launch - did you reconsider?